### PR TITLE
Introduce `builtins.groupBy` primop

### PR DESCRIPTION
**Pull request by Infinisil**
This function is very useful in nixpkgs, but its implementation in Nix
itself is rather slow due to it requiring a lot of attribute set and
list appends.

### CODE REVIEW

**Comment by edolstra** [comment]
Looks good, do you have any numbers on the performance improvement?

File: doc/manual/src/release-notes/rl-next.md [hunk 1]
```diff
@@ -5,3 +5,6 @@
 * `nix develop` now has a flag `--unpack` to run `unpackPhase`.
 
 * Lists can now be compared lexicographically using the `<` operator.
+
+* New built-in function: `builtins.groupBy`, with the same functionality as
+  nixpkgs `lib.groupBy`, but faster
```
Comment (edolstra) [inline doc/manual/src/release-notes/rl-next.md:10]:
```suggestion
  Nixpkgs' `lib.groupBy`, but faster.
```

**Comment by Infinisil** [comment]
> Looks good, do you have any numbers on the performance improvement?

Yup, see https://github.com/NixOS/nixpkgs/pull/147077#issuecomment-984909819

File: src/libexpr/primops.cc [hunk 2]
```diff
@@ -2928,6 +2928,57 @@ static RegisterPrimOp primop_partition({
...
 });
 
+static void prim_groupBy(EvalState & state, const Pos & pos, Value * * args, Value & v)
+{
+    state.forceFunction(*args[0], pos);
+    state.forceList(*args[1], pos);
+
+    ValueVectorMap attrs;
+
+    for (unsigned int n = 0; n < args[1]->listSize(); ++n) {
```
Comment (edolstra) [inline src/libexpr/primops.cc:2938]: This can be written using the list iterator now, i.e. `for (auto vElem : args[1]->listItems()) ...`.

Comment (Infinisil) [reply src/libexpr/primops.cc]: Nice, did this

**Comment by roberth** [comment]
@Infinisil @edolstra doesn't this need tests?

**Comment by gilligan** [comment]
@Infinisil Could you add tests for this? There are various tests that have been merged already and some open PR where @andir has been working on covering most primops.

I’m sure andi (or i though andi now has more practice) would also be willing to help out in case you don’t know how.

I just really really think that whenever new builtins are being added there should be *tests*. At the very minimum the code that is presented as example in the documentation.

**Comment by Infinisil** [comment]
Added one in https://github.com/NixOS/nix/pull/5786

**Pull request merged**